src/adapter/src/rbac.rs

```diff
@@ -418,6 +448,417 @@ fn ownership_err(
     }
 }
 
+/// The result of this function is a set of tuples of the form
+/// (What object the privilege is on, What privilege is required, Who must possess the privilege).
+///
+/// We use a [`BTreeSet`] because we may generate duplicate privileges while traversing through the
+/// plan and used objects.
+fn generate_required_privileges(
+    catalog: &impl SessionCatalog,
+    plan: &Plan,
+    depends_on: &Vec<GlobalId>,
+    role_id: RoleId,
+) -> BTreeSet<(ObjectId, AclMode, RoleId)> {
+    match plan {
+        Plan::CreateConnection(plan) => {
+            let mut privileges = BTreeSet::from([(
+                plan.name.qualifiers.clone().into(),
+                AclMode::CREATE,
+                role_id,
+            )]);
+            privileges.extend(generate_item_usage_privileges(catalog, depends_on, role_id));
+            privileges
+        }
+        Plan::CreateSchema(plan) => match plan.database_spec {
+            ResolvedDatabaseSpecifier::Ambient => BTreeSet::new(),
+            ResolvedDatabaseSpecifier::Id(database_id) => {
+                BTreeSet::from([(database_id.into(), AclMode::CREATE, role_id)])
+            }
+        },
+        Plan::CreateClusterReplica(plan) => {
+            BTreeSet::from([(plan.cluster_id.into(), AclMode::CREATE, role_id)])
+        }
+        Plan::CreateSource(plan) => {
+            generate_required_source_privileges(catalog, plan, depends_on, role_id)
+        }
+        Plan::CreateSources(plans) => plans
+            .iter()
+            .flat_map(|plan| {
+                // Sub-sources depend on not-yet created sources, so we need to filter those out.
+                let existing_depends_on = plan
+                    .depends_on
+                    .iter()
+                    .filter(|id| catalog.try_get_item(id).is_some())
+                    .cloned()
+                    .collect();
+                generate_required_source_privileges(
+                    catalog,
+                    &plan.plan,
+                    &existing_depends_on,
+                    role_id,
+                )
+                .into_iter()
+            })
+            .collect(),
+        Plan::CreateSecret(plan) => BTreeSet::from([(
+            plan.name.qualifiers.clone().into(),
+            AclMode::CREATE,
+            role_id,
+        )]),
+        Plan::CreateSink(plan) => {
+            let mut privileges = BTreeSet::from([(
+                plan.name.qualifiers.clone().into(),
+                AclMode::CREATE,
+                role_id,
+            )]);
+            if let SourceSinkClusterConfig::Existing { id } = &plan.cluster_config {
```

Comment:
I'm a bit scared of manually stuffing variants here and above. Does it make sense to have some named functions on `SourceSinkClusterConfig` that make the intention clearer?

Reply:
I think that makes sense, I added a function and used that. Let me know what you think.